packages/@lwc/engine-dom/src/custom-elements/create-pivot-constructor.ts

```diff
@@ -0,0 +1,462 @@
+                    awaiting.delete(this);
+                }
+            }
+        }
+        adoptedCallback(this: HTMLElement) {
+            const definition = definitionForElement.get(this);
+            definition?.adoptedCallback?.call(this);
+        }
+        attributeChangedCallback(this: HTMLElement, name: string, oldValue: any, newValue: any) {
+            const definition = definitionForElement.get(this);
+            // if both definitions are the same, then the observedAttributes is the same,
+            // but if they are different, only if the runtime definition has the attribute
+            // marked as observed, then it should invoke attributeChangedCallback.
+            if (registeredDefinition === definition || definition?.observedAttributes.has(name)) {
+                definition.attributeChangedCallback?.apply(this, [name, oldValue, newValue]);
+            }
+        }
+        static observedAttributes = [...registeredDefinition.observedAttributes];
+    }
+    return PivotCtor;
+}
+
+function getNewObservedAttributes(
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
+) {
+    const { observedAttributes, attributeChangedCallback } = instanceDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        // This instance does not need to observe any attributes, no need to patch
+        return EMPTY_SET as Set<string>;
+    }
+
+    // Natively, the attributes observed by the registered definition are going to be taken
+    // care of by the browser, only the difference between the two sets has to be taken
+    // care by the patched version.
+    return new Set(
+        [...instanceDefinition.observedAttributes].filter(
+            (x) => !registeredDefinition.observedAttributes.has(x)
+        )
+    );
+}
+
+function throwAsyncError(error: unknown) {
+    // Per native custom element behavior, errors thrown in attributeChangedCallback
+    // become unhandled async errors. We use setTimeout() instead of Promise.resolve()
+    // to make it an unhandled error rather than an unhandled rejection.
+    setTimeout(() => {
+        throw error;
+    });
+}
+
+// Helper to patch `setAttribute`/`getAttribute` to implement `attributeChangedCallback`.
+// Why is this necessary? Well basically, you can't change the `observedAttributes` after
+// a custom element is defined. So with pivots, if two classes share the same tag name,
+// and the second class observes attributes that aren't observed by the first one,
+// then those attributes can never be observed by the native `observedAttributes` system.
+// So we have to simulate it by patching `getAttribute`/`removeAttribute`. Note that
+// we only do this when absolutely necessary, though; i.e. because we've determined
+// that we aren't observing the attributes we need to.
+function patchAttributes(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
```

Comment:
Suggestion: Rename these two variable to make the concept more intuitive. I think the two have this meaning
`registeredDefinition` -> The definition to represent the constructor that was used to register during `customElements.define`
`instanceDefinition` -> The definition to represent the constructor that is passed when the pivot constructor is invoked with another constructor.

May be name it `registeredDefinition` & `pivotDefinition`? Or you could add JSDoc to the method 🙏

Reply:
Will do!